### PR TITLE
Remove contextChain from Subthoughts

### PR DESCRIPTION
**Pull request by shresthabijay**
### Changes

- Remove `contextChain` from `Subthoughts`, `Thoughts` and other related places.
- Add `appendChildPath` selector that takes parent's `path`, child's `simplePath` and returns `path` for the `child`. Pass calculated `path` to `Subthoughts` and `Thoughts`.
- Add test for `appendChildPath`.
- Update `Subthoughts` component test to use `props.path` for context view related test.
- Add new test to calculate proper `path` for a children inside context view with duplicate lexeme. Related discussion:  https://github.com/cybersemics/em/pull/878#issuecomment-717057916

### CODE REVIEW

File: src/components/__tests__/Subthoughts.js [hunk 1]
```diff
@@ -167,4 +175,54 @@ describe('context view', () => {
...
+    // enable Context View on /a/b/c/d/c
+    store.dispatch({ type: 'setCursor', path: [
+      { value: 'a', rank: 0 },
+      { value: 'b', rank: 0 },
+      { value: 'c', rank: 0 },
+      { value: 'd', rank: 0 },
+      { value: 'c', rank: 0 },
+    ] })
+
+    store.dispatch({ type: 'toggleContextView' })
```
Comment (raineorshine) [inline src/components/__tests__/Subthoughts.js:199]: Pass an array of actions to `store.dispatch`

File: src/components/__tests__/Subthoughts.js [hunk 2]
```diff
@@ -167,4 +175,54 @@ describe('context view', () => {
...
+    */
+
+    // update DOM
+    wrapper.update()
+
+    const subthoughtsFirst = wrapper
+      .find(Subthoughts)
+      .filterWhere(whereResolvedContext(['a', 'b', 'c', 'd', 'c', 'b']))
+
+    expect(subthoughtsFirst).toHaveLength(1)
```
Comment (raineorshine) [inline src/components/__tests__/Subthoughts.js:219]: Should we assert not just the length, but that it is the expected thought?

File: src/components/__tests__/Subthoughts.js [hunk 3]
```diff
@@ -167,4 +175,54 @@ describe('context view', () => {
...
+
+    const subthoughtsSecond = wrapper
+      .find(Subthoughts)
+      .filterWhere(whereResolvedContext(['a', 'b', 'c', 'd', 'c', 'd']))
+
+    expect(subthoughtsSecond).toHaveLength(1)
```
Comment (raineorshine) [inline src/components/__tests__/Subthoughts.js:225]: same

File: src/components/__tests__/Subthoughts.js [hunk 4]
```diff
@@ -167,4 +175,54 @@ describe('context view', () => {
...
+              - a.b.c (first)
+              - a.b.c.d.c (second)
+    */
+
+    // update DOM
+    wrapper.update()
+
+    const subthoughtsFirst = wrapper
+      .find(Subthoughts)
+      .filterWhere(whereResolvedContext(['a', 'b', 'c', 'd', 'c', 'b']))
```
Comment (raineorshine) [inline src/components/__tests__/Subthoughts.js:217]: Awesome! This looks great.

File: src/selectors/__tests__/getChildResolvedPath.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import getChildResolvedPath from '../getChildResolvedPath'
+import { initialState, rankThoughtsSequential, reducerFlow, pathToContext } from '../../util'
+import { Path, SimplePath } from '../../types'
+import { toggleContextView, setCursor } from '../../reducers'
+
+it('get child thoughtsResolved', async () => {
+
+  const parentThoughtsResolved = rankThoughtsSequential(['m', 'n', 'o', 'p']) as Path
+  const childSimplePath = rankThoughtsSequential(['a', 'b', 'p', 'o', 'q']) as SimplePath
```
Comment (raineorshine) [inline src/selectors/__tests__/getChildResolvedPath.ts:9]: `/m` and `/a` can't have the same rank since they are both in the root context. Not relevant for this test, but invalid states could lead to inconsistencies later on.

File: src/selectors/getChildResolvedPath.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * getChildResolved(state, ['a', 'b' , 'c'], ['a', 'b', 'c', 'd'])
+ *
+ * // returns ['m', 'n', 'o', 'p', 'q']
+ * getChildResolved(state, ['m', 'n', 'o', 'p'], ['a', 'b', 'p', 'o', 'q'])
+ *
+ * // Note: ['i', 'j', 'k'] has active context view
+ * // returns [i', 'j', 'k', 't']
+ * getChildResolved(state, ['i', 'j', 'k'], ['r', 's', 't', 'k'])
+ * */
+const getChildResolvedPath = (state: State, parentThoughtsResolved: Path, childPath: SimplePath): Path => {
```
Comment (raineorshine) [inline src/selectors/getChildResolvedPath.ts:19]: Let's call this `appendChildPath`

File: src/selectors/getChildResolvedPath.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import { SimplePath, Path } from '../types'
+import { parentOf, head, pathToContext, unroot } from '../util'
+import { State } from '../util/initialState'
+import isContextViewActive from './isContextViewActive'
+
+/** Get thoughtsResolved for a child from it's parent thoughtsResolved.
```
Comment (raineorshine) [inline src/selectors/getChildResolvedPath.ts:6]: Better description needed, e.g. Appends the head of a child `SimplePath` to a parent `Path`.

File: src/selectors/getChildResolvedPath.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * @example
+ * // returns ['a', 'b', 'c', 'd']
```
Comment (raineorshine) [inline src/selectors/getChildResolvedPath.ts:9]: Make a note here that the examples use `Contexts` only for clarity; the function actually takes `Paths`.

File: src/selectors/getChildResolvedPath.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * // returns ['m', 'n', 'o', 'p', 'q']
+ * getChildResolved(state, ['m', 'n', 'o', 'p'], ['a', 'b', 'p', 'o', 'q'])
+ *
+ * // Note: ['i', 'j', 'k'] has active context view
+ * // returns [i', 'j', 'k', 't']
+ * getChildResolved(state, ['i', 'j', 'k'], ['r', 's', 't', 'k'])
+ * */
+const getChildResolvedPath = (state: State, parentThoughtsResolved: Path, childPath: SimplePath): Path => {
+  const isParentContextViewActive = isContextViewActive(state, pathToContext(parentThoughtsResolved))
+  return unroot(parentThoughtsResolved.concat(isParentContextViewActive ? head(parentOf(childPath)) : head(childPath)))
```
Comment (raineorshine) [inline src/selectors/getChildResolvedPath.ts:21]: Prefer the spread operator over `concat` (better performance and arguably easier to read).

Comment (raineorshine) [inline src/selectors/getChildResolvedPath.ts:21]: Notice the duplication of `head` in each branch of the ternary; you should move it one level up since it is unconditional.

File: src/components/Content.tsx [hunk 10]
```diff
@@ -114,6 +114,7 @@ const Content: ContentComponent = props => {
         ? <Search />
         : <React.Fragment>
           {rootThoughtsLength === 0 ? <NewThoughtInstructions childrenLength={rootThoughtsLength} isTutorial={isTutorialLocal} /> : <Subthoughts
+            thoughtsResolved={[]}
```
Comment (raineorshine) [inline src/components/Content.tsx:117]: Note that `[]` is not a valid `Path`; a `Path` must have length of at least 1. (Unfortunately the type checker is not able to catch this.)

Comment (raineorshine) [inline src/components/Content.tsx:117]: I might suggest making `path` optional. If it is not present we can simply use `simplePath`.

File: src/components/Superscript.tsx [hunk 11]
```diff
@@ -3,11 +3,11 @@ import { connect } from 'react-redux'
 import { exists, getContexts } from '../selectors'
 import { parentOf, equalArrays, head, headValue, pathToContext, rootedParentOf } from '../util'
 import { State } from '../util/initialState'
-import { Child, Context, Index, SimplePath } from '../types'
+import { Child, Context, Index, SimplePath, Path } from '../types'
 
 interface SuperscriptProps {
   contextViews?: Index<boolean>,
-  contextChain?: SimplePath[],
+  thoughtsResolved?: Path,
```
Comment (raineorshine) [inline src/components/Superscript.tsx:10]: It doesn't look like `thoughtsResolved` is used?

File: src/selectors/__tests__/getChildResolvedPath.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  expect(pathToContext(thoughtsResolved)).toEqual(['m', 'n', 'o', 'p', 'q'])
+})
+
+it('get child thoughtsResolved when parent has active context view', async () => {
+
+  const parentThoughtsResolved = rankThoughtsSequential(['i', 'j', 'k']) as Path
+  const childSimplePath = rankThoughtsSequential(['r', 's', 't', 'k']) as SimplePath
+
+  const steps = [
+    setCursor({ path: parentThoughtsResolved }),
```
Comment (raineorshine) [inline src/selectors/__tests__/getChildResolvedPath.ts:22]: You have to import the thoughts first for this to be a valid `setCursor` call. At the moment it does not error, but we should expect this to not work.

**Review by raineorshine — CHANGES_REQUESTED** [review]

**Comment by shresthabijay** [comment]
@raineorshine I have made the necessary changes. Please review it. Thanks!

File: src/components/Subthoughts.tsx [hunk 13]
```diff
@@ -105,15 +105,17 @@ const mapStateToProps = (state: State, props: SubthoughtsProps) => {
     showHiddenThoughts,
   } = state
 
+  const resolvedPath = props.path ?? unroot(props.simplePath)
```
Comment (raineorshine) [inline src/components/Subthoughts.tsx:108]: What requires that `props.simplePath` be unrooted? You should probably move this downstream, if it's needed at all. We should only unroot paths during concatenation, otherwise we could end up with an invalid empty `Path`.

File: src/components/Subthoughts.tsx [hunk 14]
```diff
@@ -367,10 +369,12 @@ export const SubthoughtsComponent = ({
   // TODO: This getThought call looking bit ambitious to me I am commenting the previous statement please check this.
   const thought = getThought(state, headValue(simplePath))
 
+  const resolvedPath = path ?? unroot(simplePath)
```
Comment (raineorshine) [inline src/components/Subthoughts.tsx:372]: same

File: src/components/ThoughtAnnotation.tsx [hunk 15]
```diff
@@ -60,10 +60,10 @@ const mapStateToProps = (state: State, props: ThoughtAnnotationProps) => {
 
   const { cursor, cursorBeforeEdit, invalidState, editingValue, showHiddenThoughts } = state
 
-  const isEditing = equalPath(cursorBeforeEdit, props.thoughtsResolved)
+  const isEditing = equalPath(cursorBeforeEdit, props.path)
   const pathLive = isEditing
-    ? parentOf(props.path).concat(head(props.showContexts ? parentOf(cursor!) : cursor!)) as SimplePath
-    : props.path
+    ? parentOf(props.simplePath).concat(head(props.showContexts ? parentOf(cursor!) : cursor!)) as SimplePath
+    : props.simplePath
```
Comment (raineorshine) [inline src/components/ThoughtAnnotation.tsx:66]: Let's call this `simplePathLive` to avoid confusion. We should also return `simplePath: simplePathLive` instead of `path: pathLive`.

File: src/selectors/__tests__/appendChildPath.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
+import appendChildPath from '../appendChildPath'
+import { initialState, rankThoughtsSequential, reducerFlow, pathToContext } from '../../util'
+import { Path, SimplePath } from '../../types'
+import { toggleContextView, setCursor } from '../../reducers'
+
+it('get child resolved path', async () => {
+  const parentThoughtsResolved: Path = [
```
Comment (raineorshine) [inline src/selectors/__tests__/appendChildPath.ts:7]: Rename `parentThoughtsResolved` to `parentPath`

File: src/selectors/__tests__/appendChildPath.ts [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+  const childSimplePath = [
+    { value: 'a', rank: 1 },
+    { value: 'b', rank: 0 },
+    { value: 'p', rank: 0 },
+    { value: 'o', rank: 0 },
+    { value: 'q', rank: 0 },
+  ] as SimplePath
+
+  const thoughtsResolved = appendChildPath(initialState(), childSimplePath, parentThoughtsResolved)
```
Comment (raineorshine) [inline src/selectors/__tests__/appendChildPath.ts:22]: Rename `thoughtsResolved` to `path`

File: src/selectors/__tests__/appendChildPath.ts [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+  expect(pathToContext(thoughtsResolved)).toEqual(['m', 'n', 'o', 'p', 'q'])
+})
+
+it('get child resolved path when parent has active context view', async () => {
+
+  const parentThoughtsResolved: Path = [
```
Comment (raineorshine) [inline src/selectors/__tests__/appendChildPath.ts:29]: rename

File: src/selectors/appendChildPath.ts [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+ * appendChildPath(state, ['a', 'b', 'c', 'd'], ['a', 'b' , 'c'])
+ *
+ * // returns ['m', 'n', 'o', 'p', 'q']
+ * appendChildPath(state, ['a', 'b', 'p', 'o', 'q'], ['m', 'n', 'o', 'p'])
+ *
+ * // Note: ['i', 'j', 'k'] has active context view
+ * // returns [i', 'j', 'k', 't']
+ * appendChildPath(state, ['r', 's', 't', 'k'], ['i', 'j', 'k'])
+ * */
+const appendChildPath = (state: State, childPath: SimplePath, parentThoughtsResolved?: Path): Path => {
```
Comment (raineorshine) [inline src/selectors/appendChildPath.ts:21]: Rename `parentThoughtsResolved` to `parentPath`

**Review by raineorshine — CHANGES_REQUESTED** [review]

File: src/components/Subthoughts.tsx (hunk 13 above)
Comment (shresthabijay) [reply src/components/Subthoughts.tsx:108]: Got it.
